### ISSUE TITLE
App configuration: show code

### ISSUE THREAD

**Issue by VivaldoMendes** [marimo-team/marimo#2038]
### Description

Following @odanoburu's suggestion in [#1975](https://github.com/marimo-team/marimo/issues/1975) that slides would be significantly improved if code (or some code) were allowed to be displayed/visible in the presentation, I would make a similar point for the App version itself. 

Having the code visible in the Notebook version but not in the App version is almost indifferent to the one who wrote the notebook. However, if the notebook is written with someone else in mind who will interact with it during a particular task, whether the code is visible or not in the App version becomes a critical point. One crucial characteristic of computational notebooks that led to their monumental success over the last decade is, I believe, the option available to hide/display specific parts of the code. 

To the best of my knowledge, all notebooks available allow you to hide/display specific parts of the code, even in a truly app view. I am providing a good example. Consider Hex notebooks, which target a segment of users for whom the App version seems more important than the notebook version itself.  However, in Hex notebooks, the App version allows the inclusion of visible code, as well as code that is not visible.  The first image below displays the Notebook version of an example on their website, while the second one shows the App version of the same notebook. Notice that in the App version, one code cell is hidden while another is displayed. 

---
Notebook view:

![a](https://github.com/user-attachments/assets/799b7046-ca3a-46b0-b37a-560c6f0ff285)

---
App view:

<img width="566" alt="aa" src="https://github.com/user-attachments/assets/9c215640-9aff-462f-a48d-bbb6ad5f77cc">

---

If this option to hide/display code in the app version is implemented in marimo, the scope of adopters of this excellent piece of software will drastically increase. 

### Suggested solution

Allow the option to hide/display code in the app version is implemented in marimo

**Comment by mscolnick** [comment]
hi @VivaldoMendes - this is already possible today. you can do `marimo run my_app.py --include-code`

**Comment by VivaldoMendes** [comment]
Hi @mscolnick, thank you for your feedback.

I was not aware of this possibility, so I tested it. Unless I am missing some trick now, as I was before, it does not solve the problem I mentioned above. Apparently, if I load the notebook as an App (`marimo run my_notebook.py --include-code`), I can not edit the notebook. I can hide/show all code, then decide which code will be visible or hidden, but this is not a practical procedure if one uses a long notebook for teaching. 

On the other hand, if I load the notebook as a notebook (`marimo edit my-notebook.py`), I can edit the document and move from App view to Notebook view quite easily by `Ctrl + . ` , but I can not choose the cells to display and hide on a discretionary basis when the App is selected. This may look like a small detail, but it makes all the difference. 

I give an example (I continue considering the case of `marimo edit my-notebook.py`). I have a notebook that includes many markdown cells and many computational cells. The output of all markdown cells will be visible in the **_App view_** (text, equations, headers, ...), and they should be so. All markdown code (input) is hidden in the **_App view_**, and it should be so. We have the same logic in the case of computational code cells: all computational code (input) is hidden, and all the output (tables, plots, ...) is visible in the **_App view_**.  

In the first image below, I present a slice of one of my notebooks (notebook view), which shows a section number, a horizontal line, and a code cell. Even though hiding markdown code takes a lot of space in the notebook, everything looks OK. In the second image I display the same part of he notebook after I did `Ctrl + .` .  What is the problem here? The notebook looks awkward without the code cell that includes the `for loop` being visible. In fact, it seems useless. What is the small detail that avoids this: allowing this particular code cell to be always visible, no matter whether I use the notebook in the App view or the Notebook view? This possibility is available in the other notebooks, from Jupyter, Observable, Pluto, to Hex.  It is critical when the notebook is used to share information and to be used in teaching.

---

![a](https://github.com/user-attachments/assets/6ab85f90-c9e7-43de-95e0-ce8279fb8bdc)

---

![aaaa](https://github.com/user-attachments/assets/0df5c3e0-3621-4c75-b687-400129ebea3f)

---

Please do not consider this comment of mine as a criticism of the status of marimo at the moment because that is not my intention. I am just trying to highlight a particular point that, from my own experience, may allow marimo to reach many more people, particularly in education. Sorry for this long post, but I could not find a shorter way to describe my point.

Thanks

**Comment by mscolnick** [comment]
If you want the user to edit the code as well, would you "mo.ui.code_editor" help?

**Comment by mscolnick** [comment]
We could add a config to each cell "include_in_app". 

But I am afraid it could cause confusion with errors that are caused from other cells that are not included. Not to mention, the user may not have the debugging tools on the left-hand panel.

**Comment by VivaldoMendes** [comment]
I don't know what the best option is to implement such functionality. In the case of Hex, I think they have something like `Add to App builder`. My intuition tells me that it has to be implemented on a cell-by-cell basis because we may have situations where we need to hide some code cells and others where the code cells should be visible. 

My point is simple: notebooks must be **_flexible_** regarding the code cells we decide to hide and those we must keep visible (regarding markdown code, there are all the reasons to keep all code (input) hidden) in both modes: app-view and notebook-view. Without such flexibility, sooner or later, nonsense will appear in a notebook in any of those modes, as my example above indicates. This flexibility is available in all the notebooks I have used, which is a significant reason notebooks have tremendously succeeded in teaching and exchanging computational information. 

Please consider this suggestion of mine as a humble contribution to the development of marimo. I do not feel much comfortable pushing for developments that other users do not consider indispensable. I read this forum and have yet to find many contributors with the same concerns as mine.

---
An image from Hex:

![a](https://github.com/user-attachments/assets/36d079c2-5da0-426b-b503-aec63dc4bc88)

**Comment by mscolnick** [comment]
This is helpful. We can this flexibility - just need to think through the best way to not overly complicate it / confuse users into accidentally exposing / not-exposing the right stuff.

**Comment by VivaldoMendes** [comment]
That is great. To make things easier, I suggest that all input codes be hidden in the App view by default, as we have now. If someone wants some specific code cells to be visible in the app view, they have to indicate that choice to the notebook.

**Comment by patrick-kidger** [comment]
+1 for this! I was just looking for this exact functionality, thinking I'd try using Marimo to put a technical presentation (about a codebase) together.

**Comment by akshayka** [comment]
@VivaldoMendes , our next release includes a library function that lets you include the code in the output of a cell, making it visible in the app view. Importantly, as I believe you've requested, the author of the notebook can on a cell-by-cell basis choose which cells include their code in the output.

For a preview of this feature you may view https://github.com/marimo-team/marimo/pull/2342. It will be available in the next release.

**Comment by akshayka** [comment]
`mo.show_code()`, released in 0.8.16, lets you programmatically include a cell's code in its output.

See the docs for more information: https://docs.marimo.io/api/outputs.html#display-cell-code-in-marimo-s-app-views

**Comment by kakila** [comment]
`mo.show_code()` does the trick per cell without logic. In case you want to conditionally show the code of the cell, the logic:

```
(mo.show_code() if show_cellcode.value else None)
```
(where show_cellcode is a `mo.ui.switch`) is added to the shown code (correctly, as it is part of the code... although the call to `mo.show_code()` is not shown... I understand why...)

If the cell to show defines a function, say `myfunc` one can get the desired behavior. This is can be done by having another cell with the code:
```
from inspect import getsource
(mo.ui.code_editor(disabled=True, value=getsource(myfunc)) if show_cellcode.value else None)
```

But I cannot find a way to get the same for arbitrary cells, say with script code:
```
import marimo
__generated_with = "0.24.0"
app = marimo.App()

# Cell to show
@app.cell
def marimo_import():
    import marimo as mo
    return (mo,)

# Switch to hide/show code
@app.cell
def _(mo):
    show_cellcode = mo.ui.switch(label="Show code")
    show_cellcode
    return (show_cellcode,)
```
How can I show the code of the `marimo_import` cell? I tried inspecting the `app` object, but it is not available within the cells
